File: Tests/ObjC/VSS002_ClientWOTokenTests.m
//
//  VSS002_ClientWOTokenTests.m
//  VirgilSDK
//
//  Created by Oleksandr Deundiak on 2/8/17.
//  Copyright © 2017 VirgilSecurity. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <XCTest/XCTest.h>

#import "VSSTestsUtils.h"
#import "VSSTestsConst.h"
#import "Mailinator.h"
#import "MailinatorConfig.h"
#import "MEmailMetadata.h"
#import "MPart.h"
#import "MEmail.h"

/// Each request should be done less than or equal this number of seconds.
static const NSTimeInterval kEstimatedRequestCompletionTime = 8.;

@interface VSS002_ClientWOTokenTests : XCTestCase

@property (nonatomic) VSSClient *client;
@property (nonatomic) VSSCrypto *crypto;
@property (nonatomic) VSSTestsUtils *utils;
@property (nonatomic) VSSTestsConst *consts;
@property (nonatomic) Mailinator *mailinator;
@property (nonatomic) NSRegularExpression *regexp;

@end

@implementation VSS002_ClientWOTokenTests

#pragma mark - Setup

- (void)setUp {
    [super setUp];
    
    self.consts = [[VSSTestsConst alloc] init];
    
    VSSServiceConfig *config = [VSSServiceConfig defaultServiceConfig];
    self.crypto = [[VSSCrypto alloc] init];
    VSSCardValidator *validator = [[VSSCardValidator alloc] initWithCrypto:self.crypto];
    
    VSSPrivateKey *privateKey = [self.crypto importPrivateKeyFromData:[[NSData alloc] initWithBase64EncodedString:self.consts.applicationPrivateKeyBase64 options:0] withPassword:self.consts.applicationPrivateKeyPassword];
    VSSPublicKey *publicKey = [self.crypto extractPublicKeyFromPrivateKey:privateKey];
    NSData *publicKeyData = [self.crypto exportPublicKey:publicKey];
    
    XCTAssert([validator addVerifierWithId:self.consts.applicationId publicKeyData:publicKeyData]);
    
    config.cardValidator = validator;
    
    self.client = [[VSSClient alloc] initWithServiceConfig:config];
    self.utils = [[VSSTestsUtils alloc] initWithCrypto:self.crypto consts:self.consts];
    
    self.regexp = [NSRegularExpression regularExpressionWithPattern:@"Your confirmation code is.+([A-Z0-9]{6})" options:NSRegularExpressionCaseInsensitive error:nil];
    self.mailinator = [[Mailinator alloc] initWithApplicationToken:self.consts.mailinatorToken serviceUrl:[[NSURL alloc] initWithString:@"https://api.mailinator.com/api/"]];
}

- (void)tearDown {
    self.client = nil;
    self.crypto = nil;
    self.consts = nil;
    self.mailinator = nil;
    self.regexp = nil;
    self.utils = nil;
    
    [super tearDown];
}

#pragma mark - Tests
- (void)testI01_VerifyEmail {
    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Verification code should be sent to email"];
    
    NSUInteger numberOfRequests = 3;
    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
    
    NSString *identity = [self.utils generateEmail];
    
    [self.client verifyIdentity:identity identityType:@"email" extraFields:nil completion:^(NSString *actionId, NSError *error) {
        XCTAssert(error == nil);
        XCTAssert(actionId.length != 0);
        
        sleep(3);
        
        NSString *identityShort = [identity substringToIndex:[identity rangeOfString:@"@"].location];
        [self.mailinator getInbox:identityShort completionHandler:^(NSArray<MEmailMetadata *> *metadataList, NSError * error) {
            XCTAssert(error == nil);
            XCTAssert(metadataList != nil);
            XCTAssert(metadataList.count == 1);
            
            [self.mailinator getEmail:metadataList[0].mid completionHandler:^(MEmail *email, NSError *error) {
                XCTAssert(error == nil);
                XCTAssert(email != nil);
                
                MPart *bodyPart = (MPart *)email.parts[0];
                
                NSTextCheckingResult *matchResult = [self.regexp firstMatchInString:bodyPart.body options:NSMatchingReportCompletion range:NSMakeRange(0, bodyPart.body.length)];
                
                XCTAssert(matchResult != nil);
                XCTAssert(matchResult.range.location != NSNotFound);
                
                NSString *match = [bodyPart.body substringWithRange:matchResult.range];
                
                NSString *code = [match substringFromIndex:match.length - 6];
                
                XCTAssert(code.length == 6);
                
                [ex fulfill];
            }];
        }];
    }];
    
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
        if (error != nil)
            XCTFail(@"Expectation failed: %@", error);
    }];
}

- (void)testI02_ConfirmEmail {
    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Verification code should be sent to email. Validation token should be obtained"];
    
    NSUInteger numberOfRequests = 4;
    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
    
    NSString *identity = [self.utils generateEmail];
    
    [self.client verifyIdentity:identity identityType:@"email" extraFields:nil completion:^(NSString *actionId, NSError *error) {
        sleep(3);
        
        NSString *identityShort = [identity substringToIndex:[identity rangeOfString:@"@"].location];
        [self.mailinator getInbox:identityShort completionHandler:^(NSArray<MEmailMetadata *> *metadataList, NSError * error) {
            [self.mailinator getEmail:metadataList[0].mid completionHandler:^(MEmail *email, NSError *error) {
                MPart *bodyPart = (MPart *)email.parts[0];
                
                NSTextCheckingResult *matchResult = [self.regexp firstMatchInString:bodyPart.body options:NSMatchingReportCompletion range:NSMakeRange(0, bodyPart.body.length)];
                
                NSString *match = [bodyPart.body substringWithRange:matchResult.range];
                
                NSString *code = [match substringFromIndex:match.length - 6];
                
                [self.client confirmIdentityWithActionId:actionId confirmationCode:code timeToLive:3600 countToLive:12 completion:^(VSSConfirmIdentityResponse *response, NSError *error) {
                    XCTAssert(error == nil);
                    XCTAssert(response != nil);
                    XCTAssert([response.identityType isEqualToString:@"email"]);
                    XCTAssert([response.identityValue isEqualToString:identity]);
                    XCTAssert(response.validationToken.length != 0);
                    
                    [ex fulfill];
                }];
            }];
        }];
    }];
    
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
        if (error != nil)
            XCTFail(@"Expectation failed: %@", error);
    }];
}

- (void)testI03_ValidateEmail {
    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Verification code should be sent to email. Validation token should be obtained. Confirmation should pass"];
    
    NSUInteger numberOfRequests = 5;
    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
    
    NSString *identity = [self.utils generateEmail];
    
    [self.client verifyIdentity:identity identityType:@"email" extraFields:nil completion:^(NSString *actionId, NSError *error) {
        sleep(3);
        
        NSString *identityShort = [identity substringToIndex:[identity rangeOfString:@"@"].location];
        [self.mailinator getInbox:identityShort completionHandler:^(NSArray<MEmailMetadata *> *metadataList, NSError * error) {
            [self.mailinator getEmail:metadataList[0].mid completionHandler:^(MEmail *email, NSError *error) {
                MPart *bodyPart = (MPart *)email.parts[0];
                
                NSTextCheckingResult *matchResult = [self.regexp firstMatchInString:bodyPart.body options:NSMatchingReportCompletion range:NSMakeRange(0, bodyPart.body.length)];
                
                NSString *match = [bodyPart.body substringWithRange:matchResult.range];
                
                NSString *code = [match substringFromIndex:match.length - 6];
                
                [self.client confirmIdentityWithActionId:actionId confirmationCode:code timeToLive:3600 countToLive:12 completion:^(VSSConfirmIdentityResponse *response, NSError *error) {
                    [self.client validateIdentity:identity identityType:@"email" validationToken:response.validationToken completion:^(NSError *error) {
                        XCTAssert(error == nil);
                        
                        [ex fulfill];
                    }];
                }];
            }];
        }];
    }];
    
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
        if (error != nil)
            XCTFail(@"Expectation failed: %@", error);
    }];
}

- (void) testA01_CreateGlobalEmailCard {
    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Global Email Virgil Card should be created"];
    
    NSUInteger numberOfRequests = 5;
    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
    
    NSString *identity = [self.utils generateEmail];
    
    [self.client verifyIdentity:identity identityType:@"email" extraFields:nil completion:^(NSString *actionId, NSError *error) {
        sleep(3);
        
        NSString *identityShort = [identity substringToIndex:[identity rangeOfString:@"@"].location];
        [self.mailinator getInbox:identityShort completionHandler:^(NSArray<MEmailMetadata *> *metadataList, NSError * error) {
            [self.mailinator getEmail:metadataList[0].mid completionHandler:^(MEmail *email, NSError *error) {
                MPart *bodyPart = (MPart *)email.parts[0];
                
                NSTextCheckingResult *matchResult = [self.regexp firstMatchInString:bodyPart.body options:NSMatchingReportCompletion range:NSMakeRange(0, bodyPart.body.length)];
                
                NSString *match = [bodyPart.body substringWithRange:matchResult.range];
                
                NSString *code = [match substringFromIndex:match.length - 6];
                
                [self.client confirmIdentityWithActionId:actionId confirmationCode:code timeToLive:3600 countToLive:12 completion:^(VSSConfirmIdentityResponse *response, NSError *error) {
                    
                    VSSCreateGlobalCardRequest *request = [self.utils instantiateEmailCreateCardRequestWithIdentity:identity validationToken:response.validationToken keyPair:nil];
                    
                    [self.client createGlobalCardWithRequest:request completion:^(VSSCard *card, NSError *error) {
                        XCTAssert(error == nil);
                        
                        XCTAssert([self.utils checkCard:card isEqualToCreateGlobalCardRequest:request]);
                        
                        [ex fulfill];
                    }];
                }];
            }];
        }];
    }];
    
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
        if (error != nil)
            XCTFail(@"Expectation failed: %@", error);
    }];
}

- (void) testA02_RevokeGlobalEmailCard {
    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Global Email Virgil Card should be created. Global Email Virgil Card should be revoked."];
    
    NSUInteger numberOfRequests = 6;
    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
    
    NSString *identity = [self.utils generateEmail];
    
    [self.client verifyIdentity:identity identityType:@"email" extraFields:nil completion:^(NSString *actionId, NSError *error) {
        sleep(3);
        
        NSString *identityShort = [identity substringToIndex:[identity rangeOfString:@"@"].location];
        [self.mailinator getInbox:identityShort completionHandler:^(NSArray<MEmailMetadata *> *metadataList, NSError * error) {
            [self.mailinator getEmail:metadataList[0].mid completionHandler:^(MEmail *email, NSError *error) {
                MPart *bodyPart = (MPart *)email.parts[0];
                
                NSTextCheckingResult *matchResult = [self.regexp firstMatchInString:bodyPart.body options:NSMatchingReportCompletion range:NSMakeRange(0, bodyPart.body.length)];
                
                NSString *match = [bodyPart.body substringWithRange:matchResult.range];
                
                NSString *code = [match substringFromIndex:match.length - 6];
                
                [self.client confirmIdentityWithActionId:actionId confirmationCode:code timeToLive:3600 countToLive:12 completion:^(VSSConfirmIdentityResponse *response, NSError *error) {
                    
                    VSSKeyPair *keyPair = [self.crypto generateKeyPair];
                    VSSCreateGlobalCardRequest *request = [self.utils instantiateEmailCreateCardRequestWithIdentity:identity validationToken:response.validationToken keyPair:keyPair];
                    
                    [self.client createGlobalCardWithRequest:request completion:^(VSSCard *card, NSError *error) {
                        VSSRevokeGlobalCardRequest *revokeRequest = [self.utils instantiateRevokeGlobalCardForCard:card validationToken:response.validationToken withPrivateKey:keyPair.privateKey];
                        
                        [self.client revokeGlobalCardWithRequest:revokeRequest completion:^(NSError *error) {
                            XCTAssert(error == nil);
                            
                            [ex fulfill];
                        }];
                    }];
                }];
            }];
        }];
    }];
    
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
        if (error != nil)
            XCTFail(@"Expectation failed: %@", error);
    }];
}

@end

File: Tests/Swift/VSS002_ClientWOTokenTests.swift
//
//  VSS002_ClientWOTokenTests.swift
//  VirgilSDK
//
//  Created by Oleksandr Deundiak on 2/8/17.
//  Copyright © 2017 VirgilSecurity. All rights reserved.
//

import XCTest
import VirgilSDK

class VSS002_ClientWOTokenTests: XCTestCase {
    
    private var client: VSSClient!
    private var crypto: VSSCrypto!
    private var utils: VSSTestUtils!
    private var consts: VSSTestsConst!
    private var regexp: NSRegularExpression!
    private var mailinator: Mailinator!
    
    // MARK: Setup
    
    override func setUp() {
        super.setUp()
        
        self.consts = VSSTestsConst()
        self.crypto = VSSCrypto()
        
        let validator = VSSCardValidator(crypto: self.crypto)
        let privateKey = self.crypto.importPrivateKey(from: Data(base64Encoded: self.consts.applicationPrivateKeyBase64)!, withPassword: self.consts.applicationPrivateKeyPassword)!
        let publicKey = self.crypto.extractPublicKey(from: privateKey)
        let publicKeyData = self.crypto.export(publicKey)
        XCTAssert(validator.addVerifier(withId: self.consts.applicationId, publicKeyData: publicKeyData))
        let config = VSSServiceConfig.default()
        validator.useVirgilServiceVerifiers = false
        config.cardValidator = validator
        
        config.cardsServiceURL = self.consts.cardsServiceURL
        config.cardsServiceROURL = self.consts.cardsServiceROURL
        config.identityServiceURL = self.consts.identityServiceURL
        config.registrationAuthorityURL = self.consts.registrationAuthorityURL
        
        self.client = VSSClient(serviceConfig: config)
        
        self.utils = VSSTestUtils(crypto: self.crypto, consts: self.consts)
        
        self.regexp = try! NSRegularExpression(pattern: "Your confirmation code is.+([A-Z0-9]{6})", options: .caseInsensitive)
        
        self.mailinator = Mailinator(applicationToken: self.consts.mailinatorToken, serviceUrl: URL(string: "https://api.mailinator.com/api/")!)
    }
    
    override func tearDown() {
        self.client = nil
        self.crypto = nil
        self.utils = nil
        self.consts = nil
        
        super.tearDown()
    }
    
    // MARK: Tests
    func testI01_VerifyEmail() {
        let ex = self.expectation(description: "Verification code should be sent to email");
        
        let numberOfRequests = 3
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            XCTAssert(error == nil)
            XCTAssert(actionId != nil)
            XCTAssert(actionId!.lengthOfBytes(using: .utf8) != 0)
            
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                XCTAssert(code.lengthOfBytes(using: .utf8) == 6)
                
                ex.fulfill()
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testI02_ConfirmEmail() {
        let ex = self.expectation(description: "Verification code should be sent to email. Validation token should be obtained");
        
        let numberOfRequests = 4
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                self.client.confirmIdentity(withActionId: actionId!, confirmationCode: code, timeToLive: 3600, countToLive: 12) { response, error in
                    XCTAssert(error == nil)
                    XCTAssert(response != nil)
                    XCTAssert(response!.identityType == "email")
                    XCTAssert(response!.identityValue == identity)
                    XCTAssert(response!.validationToken.lengthOfBytes(using: .utf8) != 0)
                    
                    ex.fulfill()
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testI03_ValidateEmail() {
        let ex = self.expectation(description: "Verification code should be sent to email. Validation token should be obtained. Confirmation should pass");
        
        let numberOfRequests = 5
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                self.client.confirmIdentity(withActionId: actionId!, confirmationCode: code, timeToLive: 3600, countToLive: 12) { response, error in
                    self.client.validateIdentity(identity, identityType: "email", validationToken: response!.validationToken) { error in
                        XCTAssert(error == nil)
                        
                        ex.fulfill()
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testA01_CreateGlobalEmailCard() {
        let ex = self.expectation(description: "Global Email Virgil Card should be created")
        
        let numberOfRequests = 5
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                self.client.confirmIdentity(withActionId: actionId!, confirmationCode: code, timeToLive: 3600, countToLive: 12) { response, error in
                    
                    let request = self.utils.instantiateEmailCreateCardRequest(withIdentity: identity, validationToken: response!.validationToken, keyPair: nil)
                    
                    self.client.createCard(with: request) { (registeredCard, error) in
                        guard error == nil else {
                            XCTFail("Failed: " + error!.localizedDescription)
                            return
                        }
                        
                        guard let card = registeredCard else {
                            XCTFail("Card is nil")
                            return
                        }
                        
                        XCTAssert(self.utils.check(card: card, isEqualToCreateCardRequest: request))
                        
                        ex.fulfill()
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testA02_RevokeGlobalEmailCard() {
        let ex = self.expectation(description: "Global Email Virgil Card should be created. Global Email Virgil Card should be revoked.")
        
        let numberOfRequests = 6
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                self.client.confirmIdentity(withActionId: actionId!, confirmationCode: code, timeToLive: 3600, countToLive: 12) { response, error in
                    let keyPair = self.crypto.generateKeyPair()
                    let request = self.utils.instantiateEmailCreateCardRequest(withIdentity: identity, validationToken: response!.validationToken, keyPair: keyPair)
                    
                    self.client.createCard(with: request) { (registeredCard, error) in
                        let revokeRequest = self.utils.instantiateRevokeGlobalCardRequestFor(card: registeredCard!, validationToken:response!.validationToken, privateKey: keyPair.privateKey)
                        
                        self.client.revokeCardWith(revokeRequest, completion: { error in
                            guard error == nil else {
                                XCTFail("Failed: " + error!.localizedDescription)
                                return
                            }
                            
                            ex.fulfill()
                        })
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testAU01_GetChallengeMessage() {
        let ex = self.expectation(description: "Global Email Virgil Card should be created. Challenge message for this card should be received.")
        
        let numberOfRequests = 6
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                self.client.confirmIdentity(withActionId: actionId!, confirmationCode: code, timeToLive: 3600, countToLive: 12) { response, error in
                    let keyPair = self.crypto.generateKeyPair()
                    let request = self.utils.instantiateEmailCreateCardRequest(withIdentity: identity, validationToken: response!.validationToken, keyPair: keyPair)
                    
                    self.client.createCard(with: request) { (registeredCard, error) in
                        self.client.getChallengeMessageForVirgilCard(withId: registeredCard!.identifier) { response, error in
                            XCTAssert(error == nil && response != nil)
                            XCTAssert(!response!.authGrantId.isEmpty && !response!.encryptedMessage.isEmpty)
                            
                            ex.fulfill()
                        }
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testAU02_AckChallengeMessage() {
        let ex = self.expectation(description: "Global Email Virgil Card should be created. Challenge message for this card should be received. Acknowledge should return access code.")
        
        let numberOfRequests = 7
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                self.client.confirmIdentity(withActionId: actionId!, confirmationCode: code, timeToLive: 3600, countToLive: 12) { response, error in
                    let keyPair = self.crypto.generateKeyPair()
                    let request = self.utils.instantiateEmailCreateCardRequest(withIdentity: identity, validationToken: response!.validationToken, keyPair: keyPair)
                    
                    self.client.createCard(with: request) { (registeredCard, error) in
                        self.client.getChallengeMessageForVirgilCard(withId: registeredCard!.identifier) { response, error in
                            XCTAssert(error == nil)
                            XCTAssert(response != nil)
                            XCTAssert(!response!.authGrantId.isEmpty && !response!.encryptedMessage.isEmpty)
                            
                            let message = try! self.crypto.decrypt(response!.encryptedMessage, with: keyPair.privateKey)
                            
                            
                            let virgilAuthPublicKeyData = Data(base64Encoded: kVSSAuthServicePublicKeyInBase64)!
                            let virgilAuthPublicKey = self.crypto.importPublicKey(from: virgilAuthPublicKeyData)!
                            let encryptedMessage = try! self.crypto.encrypt(message, for: [virgilAuthPublicKey])
                            
                            self.client.ackChallengeMessage(withAuthGrantId: response!.authGrantId, encryptedMessage: encryptedMessage) { code, error in
                                XCTAssert(error == nil && code != nil)
                                XCTAssert(!code!.isEmpty)
                                
                                ex.fulfill()
                            }
                        }
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testAU03_ObtainAccessToken() {
        let ex = self.expectation(description: "Global Email Virgil Card should be created. Challenge message for this card should be received. Acknowledge should return access code. Token should be received.")
        
        let numberOfRequests = 8
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                self.client.confirmIdentity(withActionId: actionId!, confirmationCode: code, timeToLive: 3600, countToLive: 12) { response, error in
                    let keyPair = self.crypto.generateKeyPair()
                    let request = self.utils.instantiateEmailCreateCardRequest(withIdentity: identity, validationToken: response!.validationToken, keyPair: keyPair)
                    
                    self.client.createCard(with: request) { (registeredCard, error) in
                        self.client.getChallengeMessageForVirgilCard(withId: registeredCard!.identifier) { response, error in
                            XCTAssert(error == nil && response != nil)
                            XCTAssert(!response!.authGrantId.isEmpty && !response!.encryptedMessage.isEmpty)
                            
                            let message = try! self.crypto.decrypt(response!.encryptedMessage, with: keyPair.privateKey)
                            
                            
                            let virgilAuthPublicKeyData = Data(base64Encoded: kVSSAuthServicePublicKeyInBase64)!
                            let virgilAuthPublicKey = self.crypto.importPublicKey(from: virgilAuthPublicKeyData)!
                            let encryptedMessage = try! self.crypto.encrypt(message, for: [virgilAuthPublicKey])
                            
                            self.client.ackChallengeMessage(withAuthGrantId: response!.authGrantId, encryptedMessage: encryptedMessage) { code, error in
                                self.client.obtainAccessToken(withAccessCode: code!) { response, error in
                                    XCTAssert(error == nil)
                                    XCTAssert(response != nil)
                                    XCTAssert(!response!.refreshToken.isEmpty)
                                    XCTAssert(!response!.accessToken.isEmpty)
                                    XCTAssert(!response!.tokenType.isEmpty)
                                    XCTAssert(response!.expiresIn != 0)
                                    
                                    ex.fulfill()
                                }
                            }
                        }
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testAU04_RefreshAccessToken() {
        let ex = self.expectation(description: "Global Email Virgil Card should be created. Challenge message for this card should be received. Acknowledge should return access code. Token should be received. Token should be refreshed.")
        
        let numberOfRequests = 9
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                self.client.confirmIdentity(withActionId: actionId!, confirmationCode: code, timeToLive: 3600, countToLive: 12) { response, error in
                    let keyPair = self.crypto.generateKeyPair()
                    let request = self.utils.instantiateEmailCreateCardRequest(withIdentity: identity, validationToken: response!.validationToken, keyPair: keyPair)
                    
                    self.client.createCard(with: request) { (registeredCard, error) in
                        self.client.getChallengeMessageForVirgilCard(withId: registeredCard!.identifier) { response, error in
                            XCTAssert(error == nil && response != nil)
                            XCTAssert(!response!.authGrantId.isEmpty && !response!.encryptedMessage.isEmpty)
                            
                            let message = try! self.crypto.decrypt(response!.encryptedMessage, with: keyPair.privateKey)
                            
                            
                            let virgilAuthPublicKeyData = Data(base64Encoded: kVSSAuthServicePublicKeyInBase64)!
                            let virgilAuthPublicKey = self.crypto.importPublicKey(from: virgilAuthPublicKeyData)!
                            let encryptedMessage = try! self.crypto.encrypt(message, for: [virgilAuthPublicKey])
                            
                            self.client.ackChallengeMessage(withAuthGrantId: response!.authGrantId, encryptedMessage: encryptedMessage) { code, error in
                                self.client.obtainAccessToken(withAccessCode: code!) { response, error in
                                    self.client.refreshAccessToken(withRefreshToken: response!.refreshToken) { response, error in
                                        XCTAssert(error == nil)
                                        XCTAssert(response != nil)
                                        XCTAssert(!response!.accessToken.isEmpty)
                                        XCTAssert(response!.expiresIn != 0)
                                        
                                        ex.fulfill()
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testAU05_VerifyAccessToken() {
        let ex = self.expectation(description: "Global Email Virgil Card should be created. Challenge message for this card should be received. Acknowledge should return access code. Token should be received. Token should be verified.")
        
        let numberOfRequests = 9
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identity = self.utils.generateEmail()
        
        self.client.verifyIdentity(identity, identityType: "email", extraFields: nil) { actionId, error in
            self.utils.getConfirmationCode(identityValue: identity, mailinator: self.mailinator) { code in
                self.client.confirmIdentity(withActionId: actionId!, confirmationCode: code, timeToLive: 3600, countToLive: 12) { response, error in
                    let keyPair = self.crypto.generateKeyPair()
                    let request = self.utils.instantiateEmailCreateCardRequest(withIdentity: identity, validationToken: response!.validationToken, keyPair: keyPair)
                    
                    self.client.createCard(with: request) { (registeredCard, error) in
                        self.client.getChallengeMessageForVirgilCard(withId: registeredCard!.identifier) { response, error in
                            XCTAssert(error == nil && response != nil)
                            XCTAssert(!response!.authGrantId.isEmpty && !response!.encryptedMessage.isEmpty)
                            
                            let message = try! self.crypto.decrypt(response!.encryptedMessage, with: keyPair.privateKey)
                            
                            
                            let virgilAuthPublicKeyData = Data(base64Encoded: kVSSAuthServicePublicKeyInBase64)!
                            let virgilAuthPublicKey = self.crypto.importPublicKey(from: virgilAuthPublicKeyData)!
                            let encryptedMessage = try! self.crypto.encrypt(message, for: [virgilAuthPublicKey])
                            
                            self.client.ackChallengeMessage(withAuthGrantId: response!.authGrantId, encryptedMessage: encryptedMessage) { code, error in
                                self.client.obtainAccessToken(withAccessCode: code!) { response, error in
                                    self.client.verifyAccessToken(response!.accessToken) { cardId, error in
                                        XCTAssert(error == nil)
                                        XCTAssert(cardId! == registeredCard!.identifier)
                                        
                                        ex.fulfill()
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
}
